### PR TITLE
Localization Messages Editor

### PR DESCRIPTION
**Pull request by zadubanm02**

### CODE REVIEW

File: packages/browser-extension/source/DevToolScript/index.tsx [hunk 1]
```diff
@@ -28,14 +37,19 @@ browser.devtools.panels
...
+
+      
+
       const pathFile = msg?.fileUrl
       const path = pathFile.substring(8)
       console.log("Path", path)
       
       if (editorElement) {
-      console.log("Editor", editor)
+      //console.log("Editor", editor, "SKLOCALE", sk_SK)
```
Comment (jozef-slezak) [inline packages/browser-extension/source/DevToolScript/index.tsx:48]: please, avoid unnecessary comments

File: packages/browser-extension/source/DevToolScript/index.tsx [hunk 2]
```diff
@@ -108,6 +126,43 @@ browser.devtools.panels
       //panelWindow.respond = function (msg) {
       //  port.postMessage(msg);
       //};
+      
+      const printer = ts.createPrinter()
```
Comment (jozef-slezak) [inline packages/browser-extension/source/DevToolScript/index.tsx:130]: please, move the "locale" related code to a specific folder/file

File: packages/browser-extension/source/localization/skSK.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export default {
```
Comment (jozef-slezak) [inline packages/browser-extension/source/localization/skSK.ts:1]: I think the underscore should be used in the filename

File: packages/browser-extension/source/util/localizations.ts [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+import { ts } from "ts-morph";
+import { Project } from "ts-morph"
+import skSK from "../localization/skSK";
+
+
+export function getValuesFromLocalizationASTJSON(astLocale:SourceFile | undefined) {
+    let english:string[] = []
+    let slovak:string[] = []
+    let locales:string[] = []
+    let positionsTable:any[] = []
```
Comment (jozef-slezak) [inline packages/browser-extension/source/util/localizations.ts:11]: avoid using any - use interfaces...

File: packages/browser-extension/source/util/localizations.ts [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+          slovak = [...slovak, locale]
+        }
+      })
+      return{
+          english,
+          slovak
+      }
+}
+
+export function createTable(english:string[], slovak:string[], panelWindow:Window) {
```
Comment (jozef-slezak) [inline packages/browser-extension/source/util/localizations.ts:96]: DOM related functions should be place in a diffrent file and not mixed with AST related code.

File: packages/playground/public/index.html [hunk 6]
```diff
@@ -54,5 +54,6 @@
     <button id="compileButton">COMPILE</button>
     <button id="githubProjectButton">GitHub fetch project</button>
     <button id="githubDependenciesButton">GitHub fetch dependencies</button>
+    <buton id="svelteExample" style="border:solid 1px">Svelte Example</buton>
```
Comment (jozef-slezak) [inline packages/playground/public/index.html:57]: not related to the locale...

File: packages/browser-extension/source/util/localizations.ts [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+      ],
+      false
+    ))
+  ];
+}
+
+export function createFileFromAST(ast:any){
+  const printer = ts.createPrinter();
+  console.log("AST", ast)
+  const final = printer.printFile(ast)
```
Comment (jozef-slezak) [inline packages/browser-extension/source/util/localizations.ts:200]: Please, create som automated jest tests. I would like to see different locale file code snippets with different formatting and assertion that we preserve formatting.

File: README.md [hunk 8]
```diff
@@ -27,6 +26,10 @@ How it works:
 
 ![](/packages/browser-extension/.github/Newpage.gif)
 
+### Edit your localization files
+
+![](/packages/browser-extension/.github/Localizations.gif)
```
Comment (jozef-slezak) [inline README.md:31]: please use more general folder name
insteado of `.github` use `.media`

File: packages/browser-extension/source/localization/localizations.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    english,
+    slovak,
+    locales,
+    positionsTable,
+    englishPositionsTable
+  }
+}
+
+
+export function saveTableValuesAndParseBack(tableBody: any, positions: LocaleWithPosition[], allPositions: Position[]) {
```
Comment (jozef-slezak) [inline packages/browser-extension/source/localization/localizations.ts:40]: Never use any! Let's put here typescript interface.

File: packages/browser-extension/source/localization/localizations.ts [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+  let englishTable: string[] = []
+  let slovakTable: string[] = []
```
Comment (jozef-slezak) [inline packages/browser-extension/source/localization/localizations.ts:42]: You must not hardcode two locales.

File: packages/browser-extension/source/localization/localizations.ts [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+  let counter = 0
+  for (var r = 0, n = tableBody.rows.length; r < n; r++) {
+    for (var c = 0, m = tableBody.rows[r].cells.length; c < m; c++) {
```
Comment (jozef-slezak) [inline packages/browser-extension/source/localization/localizations.ts:45]: Strange data struscutre used - just a simple array. Use objects inside arrays to take advantage of compiler validations.

File: packages/browser-extension/source/localization/localizations.ts [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        counter = counter + 1
+      }
+    }
+  }
+  console.log('New Positions', positions, "New All Positions", allPositions)
+  return { englishTable, slovakTable, positions, allPositions }
+
+}
+
+export const extractWord = (code: string, start: number, end: number) => {
```
Comment (jozef-slezak) [inline packages/browser-extension/source/localization/localizations.ts:65]: strange naming - you are not really extracting  a single word...

File: packages/browser-extension/source/localization/localizationInterfaces.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+export interface LocaleWithPosition {
+    text: string;
```
Comment (jozef-slezak) [inline packages/browser-extension/source/localization/localizationInterfaces.ts:2]: please use better naming (this is too general naming)

check formatjs/react-intl documentation and inspire by bettern naming (see the <FormattedMessage attributes and use theme here)

File: packages/browser-extension/source/localization/localizationInterfaces.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+    position: number;
+}
+
+export interface Position {
+    text: string;
```
Comment (jozef-slezak) [inline packages/browser-extension/source/localization/localizationInterfaces.ts:7]: please use better naming (this is too general naming)

check formatjs/react-intl documentation and inspire by bettern naming (see the <FormattedMessage attributes and use theme here)

File: packages/browser-extension/source/localization/localizationInterfaces.ts [hunk 15]
```diff
@@ -0,0 +1,10 @@
+export interface LocaleWithPosition {
```
Comment (jozef-slezak) [inline packages/browser-extension/source/localization/localizationInterfaces.ts:1]: Strange interface name. Maybe `Message` or `LocalizedMessage` would be suitable


interface LocalizedMessage {
   id
   value
}

interface SourcePosition { // this interface is more general we might import it in here
   start
   end
}

type alias XY = LocalizedMessage | SourcePosition

File: packages/browser-extension/source/localization/localizations.ts [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+      positionsTable = [...positionsTable, { text: property.name.text, start: property.name.pos, end: property.name.end },
+      { text: property.initializer.text, start: property.initializer.pos, end: property.initializer.end }]
+      englishPositionsTable = [...englishPositionsTable, { text: property.name.text, position: property.name.pos }]
+      english = [...english, property.name.text]
+      slovak = [...slovak, property.initializer.text]
+    })
+  })
+  console.log("POsitions", positionsTable, "English positions", englishPositionsTable)
+  return {
+    english,
```
Comment (jozef-slezak) [inline packages/browser-extension/source/localization/localizations.ts:31]: do not hardcode two locales

File: packages/browser-extension/source/localization/localizations.ts [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+      locales = [...locales, property.name.text]
+      locales = [...locales, property.initializer.text]
+      positionsTable = [...positionsTable, { text: property.name.text, start: property.name.pos, end: property.name.end },
+      { text: property.initializer.text, start: property.initializer.pos, end: property.initializer.end }]
+      englishPositionsTable = [...englishPositionsTable, { text: property.name.text, position: property.name.pos }]
+      english = [...english, property.name.text]
+      slovak = [...slovak, property.initializer.text]
+    })
+  })
+  console.log("POsitions", positionsTable, "English positions", englishPositionsTable)
```
Comment (jozef-slezak) [inline packages/browser-extension/source/localization/localizations.ts:29]: awoid strange console logs (if needed use trace)

File: packages/browser-extension/source/localization/localizations.test.ts [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+import sk_SK from "./sk_SK"
+
+const expectedAST = createAst(JSON.stringify(sk_SK), ScriptTarget.ESNext, ScriptKind.JSON)
+const { english, slovak, locales, positionsTable } = getValuesFromLocalizationASTJSON(expectedAST)
+const originalWords = extractAllStrings(positionsTable)
+const positionsConst = positionsTable
+const positionLength = positionsTable.length
+//change this array for making changes of final result json
+// i.e. change text property
+const positionsChanged = [{ "text": "year", "start": 1, "end": 7 }, { "text": "xxxx", "start": 8, "end": 13 }, { "text": "month", "start": 14, "end": 21 }, { "text": "mesiac", "start": 22, "end": 30 }, { "text": "day", "start": 31, "end": 36 }, { "text": "den", "start": 37, "end": 42 }, { "text": "minute", "start": 43, "end": 51 }, { "text": "xxx", "start": 52, "end": 60 }, { "text": "second", "start": 61, "end": 69 }, { "text": "sekunda", "start": 70, "end": 79 }, { "text": "new", "start": 80, "end": 85 }, { "text": "novy", "start": 86, "end": 92 }, { "text": "old", "start": 93, "end": 98 }, { "text": "starky", "start": 99, "end": 107 }, { "text": "super", "start": 108, "end": 115 }, { "text": "super", "start": 116, "end": 123 }]
```
Comment (jozef-slezak) [inline packages/browser-extension/source/localization/localizations.test.ts:14]: make the tests user/developer friendlly
just compare two ASTs in the assertion

File: packages/browser-extension/source/DevToolScript/index.tsx [hunk 19]
```diff
@@ -109,6 +123,52 @@ browser.devtools.panels
       //panelWindow.respond = function (msg) {
       //  port.postMessage(msg);
       //};
+
+      console.log("sk_SK stringify",JSON.stringify(sk_SK))
+      
+      const loadedFile = await loadFileFromReactProject('/Users/michalzaduban/Desktop/talentsbase/src/localizations/en_EN.json')
+      //@ts-ignore
```
Comment (jozef-slezak) [inline packages/browser-extension/source/DevToolScript/index.tsx:130]: why do we have `//@ts-ignore` here?

File: packages/browser-extension/source/localization/en_EN.ts [hunk 20]
```diff
@@ -0,0 +1,10 @@
+export default {
```
Comment (jozef-slezak) [inline packages/browser-extension/source/localization/en_EN.ts:1]: we should not have this file here

File: packages/browser-extension/source/localization/localizations.ts [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+      console.log("")
+    } else {
+      const before = enSourceCode.substring(0, originalMessages[i].enEN.position.pos + 1)
+      const after = enSourceCode.substring(originalMessages[i].enEN.position.end - 1)
+      enSourceCode = before + changedMessages[i].enEN.value + after
+    }
+  }
+  return { skSourceCode, enSourceCode }
+}
+
```
Comment (jozef-slezak) [inline packages/browser-extension/source/localization/localizations.ts:117]: write just one terminating line...

File: packages/browser-extension/source/localization/en_EN.ts (hunk 20 above)
Comment (zadubanm02) [reply packages/browser-extension/source/localization/en_EN.ts:1]: this file is related to this project, en_EN.json is related to react app project

File: packages/browser-extension/source/DevToolScript/index.tsx (hunk 19 above)
Comment (zadubanm02) [reply packages/browser-extension/source/DevToolScript/index.tsx:130]: bc loadedFile is string | void and function createAst accepts string as the first argument

File: packages/browser-extension/source/DevToolScript/index.tsx [hunk 22]
```diff
@@ -157,6 +159,25 @@ browser.devtools.panels
...
+      saveDynamic?.addEventListener('click', async ()=>{
+        //@ts-ignore
+      const savedDynamicLocales = saveAllLocalesFromTable(dynamicTableBody, dynamicLocales)
+      console.log("ALl messages", savedDynamicLocales, "Original", dynamicLocales)
+      if(savedDynamicLocales){
+        const changedFiles = await updateFiles(filesFromDirectory, savedDynamicLocales, originalDynamic)
+        console.log("CHANGED FILES", changedFiles)
+        //@ts-ignore
+        sendUpdatedFiles(changedFiles, fileNames)
+        //const {skSourceCode, enSourceCode} = changeAllFiles(JSON.stringify(sk_SK),JSON.stringify(en_EN),allMessages, originalMulti)
```
Comment (jozef-slezak) [inline packages/browser-extension/source/DevToolScript/index.tsx:173]: please remove unused code

File: packages/browser-extension/source/DevToolScript/index.tsx [hunk 23]
```diff
@@ -138,12 +138,14 @@ browser.devtools.panels
...
       const dynamicLocales = createDynamicLocales(temporary)
+      const originalDynamic = JSON.parse(JSON.stringify(dynamicLocales));
       //@ts-ignore
       const loadedLocale = createAst(loadedFile,ScriptTarget.ESNext,ScriptKind.JSON )
       const sourceCodes = [sk_SK, en_EN]
       console.log("SOurce codes", sourceCodes)
       const all = combineLocales(sourceCodes)
-      
+      //@ts-ignore
+      const dynamicTableBody:HTMLTableElement= panelWindow.document.getElementById('dynamic-tableBody')
```
Comment (jozef-slezak) [inline packages/browser-extension/source/DevToolScript/index.tsx:148]: remove `//@ts-ignore`
`const dynamicTableBody = panelWindow.document.getElementById('dynamic-tableBody') as HTMLTableElement` at all simmilar places

File: packages/browser-extension/source/DevToolScript/index.tsx [hunk 24]
```diff
@@ -138,12 +138,14 @@ browser.devtools.panels
       console.log("Temporary", temporary)
```
Comment (jozef-slezak) [inline packages/browser-extension/source/DevToolScript/index.tsx:138]: please remove development comments

**Pull request merged**